test_qmk_compiler.py

```diff
@@ -0,0 +1,337 @@
+"""Test suite for qmk_compiler.
+
+This file tests most of the functionality in qmk_compiler. Given the nature
+of qmk_compiler most of the functionality depends on qmk_firmware being
+checked out and available. To satisfy this requirement
+`test_0000_checkout_qmk_skip_cache()` must be tested first, and
+`test_9999_teardown()` must be tested last. It would be better if this were
+setup as a test fixture, and some day it will be. Setting up a proper test
+fixture here is a non-trivial task and will take work that the author has
+not had time to put in yet.
+"""
+import filecmp
+import os
+import os.path
+import re
+import shutil
+from tempfile import mkstemp
+
+import pytest
+
+import update_kb_redis
+import qmk_commands
+
+
+############################################################################
+# Setup Environment                                                        #
+############################################################################
+
+
+def test_0000_checkout_qmk_skip_cache():
+    """Make sure that we successfully git clone qmk_firmware and generate the version.txt hash.
+    """
+    qmk_commands.checkout_qmk(skip_cache=True)
+    assert os.path.exists('qmk_firmware/version.txt')
+
+
+############################################################################
+# Begin Tests                                                              #
+############################################################################
+
+
+# Source code retrieval testing. Make sure we're storing and fetching the
+# correct source every way we can. At least test_0001 must be run before any
+# other test in this file or they will all fail.
+def test_0001_fetch_source_qmk():
+    """Make sure that the qmk_firmware.zip the previous test uploaded is the same as the one on disk.
+    """
+    os.rename('qmk_firmware.zip', 'qmk_firmware_cloned.zip')
+    qmk_commands.fetch_source('qmk_firmware')
+    assert filecmp.cmp('qmk_firmware_cloned.zip', 'qmk_firmware.zip')
+    os.remove('qmk_firmware_cloned.zip')
+    os.remove('qmk_firmware.zip')
+
+
+def test_0002_checkout_qmk_cache():
+    """Make sure that we fetch the cache from QMK and don't clone from git.
+    """
+    git_hash = open('qmk_firmware/version.txt', 'r').read()
```

Comment:
`open()` is a python built-in. Technically it could throw exceptions not related to the test, but these are functional tests and not unit tests. We can mock interfaces for unit tests at a later time.